common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -460,47 +458,45 @@ public String toString() {
             "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation("
     };
 
-    static String newRecord(Object hint, int recordsToSkip) {
-        StringBuilder buf = new StringBuilder(4096);
+    private static final class Record extends Throwable {
+        private final String hintString;
+        private Record next;
 
-        // Append the hint first if available.
-        if (hint != null) {
-            buf.append("\tHint: ");
-            // Prefer a hint string to a simple string form.
-            if (hint instanceof ResourceLeakHint) {
-                buf.append(((ResourceLeakHint) hint).toHintString());
-            } else {
-                buf.append(hint);
-            }
-            buf.append(NEWLINE);
+        Record(Object hint) {
+            // This needs to be generated even if toString() is never called as it may change later on.
+            hintString = hint instanceof ResourceLeakHint ? ((ResourceLeakHint) hint).toHintString() : hint.toString();
         }
 
-        // Append the stack trace.
-        StackTraceElement[] array = new Throwable().getStackTrace();
-        for (StackTraceElement e: array) {
-            if (recordsToSkip > 0) {
-                recordsToSkip --;
-            } else {
-                String estr = e.toString();
+        Record() {
+           hintString = null;
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder buf = new StringBuilder(4096);
+            if (hintString != null) {
+                buf.append("\tHint: ").append(hintString).append(NEWLINE);
+            }
+
+            // Append the stack trace.
+            StackTraceElement[] array = getStackTrace();
+            // Skip the first three elements.
+            outer: for (int i = 3; i < array.length; i++) {
+                String estr = array[i].toString();
```

Comment:
nit: if you remove the `(` characters from the elements in `STACK_TRACE_ELEMENT_EXCLUSIONS`, you don't have to call `toString` here.  You can just call `getMethodName`.

Reply:
good idea.

Reply:
actually this is not quite true as I also need to take the classname into account. That said I think we can optimise it.